File: projects/voice-inbox/src/system-status.ts
/**
 * System/thread/queue status for the operator-facing dashboard view
 * (GET /api/v1/system/status). Reads of ~/.pa/topic-threads, worker-pids,
 * blackboard.json etc. plus psutil-derived CPU/mem/disk/uptime are ALL owned
 * by scripts/system_status.py (see that file's docstring for why: reusing
 * the already-working psutil logic beats re-implementing Windows
 * disk/process-memory reads in Node). This module just spawns it and caches
 * the result so a 2-3s UI poll from multiple tabs doesn't fork a process per
 * request.
 *
 * Serving semantics (2026-09-13, vi-2d9444e29d52 recurrence): the collector
 * can take 10-25s on a degraded machine (the ~33k-file pa-dir walk alone was
 * measured at 11.5s), far past any per-request budget — the pre-fix 8s spawn
 * timeout made every poll whose dir-size cache had lapsed return 500 and the
 * PWA showed "unavailable". So a fresh collection NEVER blocks a request that
 * already has data: serve the last good snapshot immediately (stale-while-
 * revalidate), refresh in the background, dedup concurrent callers onto one
 * in-flight spawn. Only a COLD cache (no snapshot yet, e.g. first poll after
 * a server restart) awaits the collection, up to TIMEOUT_MS. A snapshot older
 * than MAX_STALE_MS is not served — a collector that has been broken for that
 * long surfaces as an honest 500 instead of stale numbers looking live. A
 * keep-warm timer additionally refreshes the snapshot every KEEP_WARM_INTERVAL_MS
 * once any request has armed it, so a dashboard click after a long idle period
 * is served from a recently refreshed snapshot instead of gambling on a cold
 * collection.
 */
import { spawn } from 'node:child_process';
import { join } from 'node:path';

export interface SystemStatus {
  generated_at: number;
  threads: {
    running_count: number;
    queued_count: number;
    running: Array<{ topic: string; id: string; title: string; updatedAt: string }>;
    queued: Array<{ topic: string; id: string; title: string; updatedAt: string }>;
    by_status: Record<string, number>;
    active_topic_count: number;
    total_topic_count: number;
  };
  workers: {
    processes: Array<{ pid: number | null; alive: boolean; worker: string | null; skill: string | null; startedAt: string | null; rss_bytes: number }>;
    alive_count: number;
    slot_used: number;
    slot_ceiling: number;
    slot_ceiling_derived: boolean;
  };
  health: { bot_pid: number | null; bot_alive: boolean; catchup_pid: number | null; catchup_alive: boolean };
  system: {
    cpu_percent: number;
    cpu_count: number | null;
    mem_total: number;
    mem_used: number;
    mem_percent: number;
    disks: Record<string, { total: number; used: number; free: number; percent: number } | null>;
    boot_time: number;
    uptime_secs: number;
  };
  assistant_mem_bytes: number;
  pa_dir_size: { bytes: number; files: number; computed_at: number };
}

export const CACHE_TTL_MS = 10_000;
export const MAX_STALE_MS = 10 * 60_000;
const TIMEOUT_MS = 60_000;
const KEEP_WARM_INTERVAL_MS = 30_000;

let lastGood: { data: SystemStatus; at: number } | null = null;
let inflight: Promise<SystemStatus> | null = null;
let keepWarmTimer: NodeJS.Timeout | null = null;
let keepWarmIntervalMs = KEEP_WARM_INTERVAL_MS;
let keepWarmCtx: {
  repoRoot: string;
  runFn: (repoRoot: string) => Promise<SystemStatus>;
  nowFn: () => number;
} | null = null;

/**
 * Test seam: the runner and clock are injectable and the module-level
 * snapshot/in-flight state is resettable (see src/tests/system-status.test.ts).
 */
export function __resetSystemStatusForTests(): void {
  lastGood = null;
  inflight = null;
  if (keepWarmTimer) {
    clearInterval(keepWarmTimer);
    keepWarmTimer = null;
  }
  keepWarmCtx = null;
  keepWarmIntervalMs = KEEP_WARM_INTERVAL_MS;
}

/** Test seam: shrink the keep-warm period; re-arms the timer if it is running. */
export function __setKeepWarmIntervalMsForTests(ms: number): void {
  keepWarmIntervalMs = ms;
  if (keepWarmTimer) {
    clearInterval(keepWarmTimer);
    keepWarmTimer = null;
    if (keepWarmCtx) keepWarmTimer = armKeepWarm();
  }
}

function startRefresh(
  repoRoot: string,
  runFn: (repoRoot: string) => Promise<SystemStatus>,
  nowFn: () => number,
): Promise<SystemStatus> {
  inflight = runFn(repoRoot)
    .then((data) => {
      lastGood = { data, at: nowFn() };
      return data;
    })
    .finally(() => {
      inflight = null;
    });
  // A rejected refresh must not become an unhandled rejection riding the
  // shared inflight promise: cold-cache callers await it (they handle the
  // rejection), but warm-cache and keep-warm callers only touch it fire-and-forget.
  inflight.catch(() => {});
  return inflight;
}

export function getSystemStatus(
  repoRoot: string,
  runFn: (repoRoot: string) => Promise<SystemStatus> = runScript,
  nowFn: () => number = Date.now,
): Promise<SystemStatus> {
  keepWarmCtx = { repoRoot, runFn, nowFn };
  if (!keepWarmTimer) keepWarmTimer = armKeepWarm();

  if (lastGood && nowFn() - lastGood.at < CACHE_TTL_MS) return Promise.resolve(lastGood.data);

  const pending = inflight ?? startRefresh(repoRoot, runFn, nowFn);

  if (lastGood && nowFn() - lastGood.at < MAX_STALE_MS) return Promise.resolve(lastGood.data);
  return pending;
}

function armKeepWarm(): NodeJS.Timeout {
  const timer = setInterval(() => {
    keepWarmTick();
  }, keepWarmIntervalMs);
  timer.unref(); // never hold the process (or a test run) open by itself
  return timer;
}

/**
 * One background tick: refresh when the snapshot is not fresh and nothing is
 * already in flight. Shares the request path's startRefresh/lastGood plumbing
 * (one pathway), so a concurrent cold caller that grabs this inflight still
 * receives a real SystemStatus.
 */
function keepWarmTick(): void {
  const ctx = keepWarmCtx;
  if (!ctx || inflight) return;
  if (lastGood && ctx.nowFn() - lastGood.at < CACHE_TTL_MS) return;
  startRefresh(ctx.repoRoot, ctx.runFn, ctx.nowFn);
}

function runScript(repoRoot: string): Promise<SystemStatus> {
  return new Promise((resolvePromise, reject) => {
    // repoRoot is the REPO root (repo-root.ts, AI-236 convention) — the script
    // lives at <repo>/projects/voice-inbox/scripts, not <repo>/voice-inbox.
    const scriptPath = join(repoRoot, 'projects', 'voice-inbox', 'scripts', 'system_status.py');
    let child;
    try {
      child = spawn('python', [scriptPath], { windowsHide: true, stdio: ['ignore', 'pipe', 'pipe'] });
    } catch (err) {
      reject(err instanceof Error ? err : new Error(String(err)));
      return;
    }
    let stdout = '';
    let settled = false;
    const done = (fn: () => void): void => {
      if (settled) return;
      settled = true;
      fn();
    };
    const timer = setTimeout(() => {
      try {
        child.kill();
      } catch {
        /* best effort */
      }
      done(() => reject(new Error('system_status.py timed out')));
    }, TIMEOUT_MS);
    child.stdout?.on('data', (chunk: Buffer) => {
      stdout += chunk.toString();
    });
    child.on('error', (err) => {
      clearTimeout(timer);
      done(() => reject(err));
    });
    child.on('close', (code) => {
      clearTimeout(timer);
      if (code !== 0) {
        done(() => reject(new Error(`system_status.py exited ${code}`)));
        return;
      }
      try {
        const parsed = JSON.parse(stdout.trim()) as SystemStatus;
        done(() => resolvePromise(parsed));
      } catch (err) {
        done(() => reject(err instanceof Error ? err : new Error(String(err))));
      }
    });
  });
}

File: projects/voice-inbox/scripts/system_status.py
"""
One-shot system/thread/queue status reader for the PA (Personal Assistant)
system, invoked by voice-inbox's `src/system-status.ts` (GET
/api/v1/system/status). Ported from the standalone `system-dashboard/server.py`
dashboard: reads existing PA state files directly (no new state format, no IPC
into the running bot process):

  ~/.pa/topic-threads/*.json   -> per-topic orchestrator thread records
                                   (status: running/queued/done/failed/cancelled)
  ~/.pa/worker-pids/*.json     -> live spawned CLI worker processes (claude/codex/agy/zclaude)
  ~/.pa/blackboard.json        -> active locks, incl. worker-slot-* (global concurrency slots)
  ~/.pa/telegram-bot.lock      -> bot process pid
  ~/.pa/catchup-loop.lock      -> pa catchup loop pid

Unlike the standalone dashboard (a `ThreadingHTTPServer` polled by a browser),
this script runs once per invocation and exits, printing a single JSON line
on stdout. `system-status.ts` spawns it and caches the result briefly so
rapid UI polls don't each fork a process.

Run: python system_status.py
"""

import json
import os
import tempfile
import time

import psutil

PA_HOME = os.path.expanduser("~/.pa")

DIR_SIZE_CACHE_PATH = os.path.join(tempfile.gettempdir(), "pa_dashboard_dirsize_cache.json")
# The walk is the collector's dominant cost: ~1s over ~33k files on a healthy
# machine, measured 11.5s on a degraded one (2026-09-13, uptime 2.7d) — the
# 60s TTL of the pre-fix era made a cold walk blow the server-side spawn
# timeout on roughly every other poll. 300s: an informational size gauge can
# be 5 minutes stale; the server (system-status.ts) serves last-good while a
# refresh runs, so a cold walk no longer blocks any response.
DIR_SIZE_REFRESH_SECS = 300

# ---------------------------------------------------------------------------
# PA state readers
# ---------------------------------------------------------------------------


def _read_json(path, default=None):
    try:
        with open(path, "r", encoding="utf-8") as fh:
            return json.load(fh)
    except (OSError, json.JSONDecodeError):
        return default


def _read_pid_file(path):
    try:
        with open(path, "r", encoding="utf-8") as fh:
            return int(fh.read().strip())
    except (OSError, ValueError):
        return None


def get_threads_and_queue():
    """Sum orchestrator thread statuses across every Telegram topic's store file."""
    topic_dir = os.path.join(PA_HOME, "topic-threads")
    running = []
    queued = []
    active_topics = set()
    total_by_status = {}
    try:
        filenames = os.listdir(topic_dir)
    except OSError:
        filenames = []

    for fname in filenames:
        if not fname.endswith(".json"):
            continue
        topic_key = fname[:-5]
        data = _read_json(os.path.join(topic_dir, fname), default={})
        if not isinstance(data, dict):
            continue
        for tid, rec in data.items():
            if not isinstance(rec, dict):
                continue
            status = rec.get("status", "unknown")
            total_by_status[status] = total_by_status.get(status, 0) + 1
            entry = {
                "topic": topic_key,
                "id": tid,
                "title": rec.get("title", ""),
                "updatedAt": rec.get("updatedAt", ""),
            }
            if status == "running":
                running.append(entry)
                active_topics.add(topic_key)
            elif status == "queued":
                queued.append(entry)
                active_topics.add(topic_key)

    return {
        "running": running,
        "queued": queued,
        "by_status": total_by_status,
        "active_topic_count": len(active_topics),
        "total_topic_count": len(filenames),
    }


def get_worker_processes():
    """Live spawned CLI workers (claude/codex/agy/zclaude), one file per PID."""
    wp_dir = os.path.join(PA_HOME, "worker-pids")
    workers = []
    try:
        filenames = os.listdir(wp_dir)
    except OSError:
        filenames = []

    for fname in filenames:
        if not fname.endswith(".json"):
            continue
        rec = _read_json(os.path.join(wp_dir, fname), default=None)
        if not isinstance(rec, dict):
            continue
        pid = rec.get("pid")
        alive = isinstance(pid, int) and psutil.pid_exists(pid)
        rss = 0
        if alive:
            for check_pid in [pid] + list(rec.get("descendants") or []):
                try:
                    rss += psutil.Process(check_pid).memory_info().rss
                except (psutil.NoSuchProcess, psutil.AccessDenied):
                    pass
        workers.append(
            {
                "pid": pid,
                "alive": alive,
                "worker": rec.get("worker"),
                "skill": rec.get("skill"),
                "startedAt": rec.get("startedAt"),
                "rss_bytes": rss,
            }
        )
    return workers


def get_worker_slot_capacity():
    """Global concurrency ceiling, mirroring pa/src/lib/dynamic-slots.ts."""
    override = os.environ.get("PA_MAX_CONCURRENT_WORKERS")
    if override is not None:
        try:
            ceiling = int(override)
            derived = False
        except ValueError:
            ceiling = max(4, min(64, psutil.cpu_count() * 4))
            derived = True
    else:
        ceiling = max(4, min(64, psutil.cpu_count() * 4))
        derived = True

    blackboard = _read_json(os.path.join(PA_HOME, "blackboard.json"), default={})
    locks = blackboard.get("active_locks", []) if isinstance(blackboard, dict) else []
    used = sum(1 for lock in locks if str(lock.get("resource", "")).startswith("worker-slot-"))

    return {"used": used, "ceiling": ceiling, "derived": derived}


def get_process_health():
    bot_pid = _read_pid_file(os.path.join(PA_HOME, "telegram-bot.lock"))
    catchup_pid = _read_pid_file(os.path.join(PA_HOME, "catchup-loop.lock"))
    return {
        "bot_pid": bot_pid,
        "bot_alive": bool(bot_pid and psutil.pid_exists(bot_pid)),
        "catchup_pid": catchup_pid,
        "catchup_alive": bool(catchup_pid and psutil.pid_exists(catchup_pid)),
    }


def get_system_stats():
    vm = psutil.virtual_memory()
    disks = {}
    for drive in ["C:/", "D:/"]:
        try:
            du = psutil.disk_usage(drive)
            disks[drive] = {"total": du.total, "used": du.used, "free": du.free, "percent": du.percent}
        except OSError:
            disks[drive] = None
    return {
        # A single fresh-process call with a real interval is accurate here —
        # the background-sampler-thread trick in the standalone dashboard only
        # existed to work around ThreadingHTTPServer handing each request a
        # new thread identity (psutil.cpu_percent(interval=None) tracks "time
        # since last call" per calling thread). Irrelevant for a one-shot script.
        "cpu_percent": psutil.cpu_percent(interval=0.3),
        "cpu_count": psutil.cpu_count(),
        "mem_total": vm.total,
        "mem_used": vm.used,
        "mem_percent": vm.percent,
        "disks": disks,
        "boot_time": psutil.boot_time(),
        "uptime_secs": time.time() - psutil.boot_time(),
    }


def get_assistant_memory_bytes(bot_pid, catchup_pid, worker_procs):
    total = 0
    seen = set()
    for pid in [bot_pid, catchup_pid]:
        if pid and psutil.pid_exists(pid) and pid not in seen:
            seen.add(pid)
            try:
                total += psutil.Process(pid).memory_info().rss
            except (psutil.NoSuchProcess, psutil.AccessDenied):
                pass
    for w in worker_procs:
        total += w["rss_bytes"]
    return total


def _pa_dir_size():
    """File-based cache (this script has no long-lived background thread to
    hold an in-memory one): repeated invocations within DIR_SIZE_REFRESH_SECS
    skip the directory walk over ~33k files (seconds, not ~1s, on a degraded
    machine — see DIR_SIZE_REFRESH_SECS)."""
    cached = _read_json(DIR_SIZE_CACHE_PATH, default=None)
    now = time.time()
    if isinstance(cached, dict) and now - cached.get("computed_at", 0) < DIR_SIZE_REFRESH_SECS:
        return cached
    total = 0
    count = 0
    for root, _dirs, files in os.walk(PA_HOME):
        for f in files:
            try:
                total += os.path.getsize(os.path.join(root, f))
                count += 1
            except OSError:
                pass
    result = {"bytes": total, "files": count, "computed_at": now}
    try:
        with open(DIR_SIZE_CACHE_PATH, "w", encoding="utf-8") as fh:
            json.dump(result, fh)
    except OSError:
        pass
    return result


def build_status():
    threads = get_threads_and_queue()
    workers = get_worker_processes()
    slots = get_worker_slot_capacity()
    health = get_process_health()
    system = get_system_stats()
    assistant_mem = get_assistant_memory_bytes(health["bot_pid"], health["catchup_pid"], workers)

    return {
        "generated_at": time.time(),
        "threads": {
            "running_count": len(threads["running"]),
            "queued_count": len(threads["queued"]),
            "running": threads["running"][:25],
            "queued": threads["queued"][:25],
            "by_status": threads["by_status"],
            "active_topic_count": threads["active_topic_count"],
            "total_topic_count": threads["total_topic_count"],
        },
        "workers": {
            "processes": workers,
            "alive_count": sum(1 for w in workers if w["alive"]),
            "slot_used": slots["used"],
            "slot_ceiling": slots["ceiling"],
            "slot_ceiling_derived": slots["derived"],
        },
        "health": health,
        "system": system,
        "assistant_mem_bytes": assistant_mem,
        "pa_dir_size": _pa_dir_size(),
    }


if __name__ == "__main__":
    print(json.dumps(build_status()))
